### ISSUE TITLE
Error for not resolved inheritance in generics that should pass

### ISSUE THREAD

**Issue by stekycz** [facebook/flow#7953]
**Flow version:** >=v0.66.0

## Code example

```js
class Parent<K> {}

class Child<K> extends Parent<K> {}

const func = (): Parent<Parent<string>> => {
  return (new Child(): Child<Child<string>>);
};
```

[Try-Flow](https://flow.org/try/#0MYGwhgzhAEAKYCcCmA7ALgHgNID5oG8BfAKGNEhgGEALASxABNs8kAPNVBmeZdZgkmQD2KCGmgAzAK4pg0ALzQAFAEoAXHESpMPbRjEJaKAOY488vPmLRoyNFIQplKJAHdoNeg1UbPjDH5MBkamOCoA3MSEkUA)

## Expected behavior

```
No errors!
```

## Actual behavior

```
Cannot return `new Child()` because `Child` [1] is incompatible with `Parent` [2] in type argument `K` [3].
```

-----

**Note:** The original issue we have found was in a function that prepares a `Map` with `Array` values. The array and the map must be writable inside the function but we want them both read-only once it leaves the function.

**Comment by goodmind** [comment]
```js
class Parent<+K> {}

class Child<K> extends Parent<K> {}

const func = (): Parent<Parent<string>> => {
  return (new Child(): Child<Child<string>>);
};
```
This works

**Comment by stekycz** [comment]
The original code looks like

```js
const func = (): $ReadOnlyMap<string, $ReadOnlyArray<string>> => {
  const map: Map<string, string[]> = new Map();

  // code that initializes the map and its values

  return map;
};
```

So I guess it is failing because `V` type in `$ReadOnlyMap` does not use `+`.

See https://github.com/facebook/flow/blob/64064e93639fe9c03d8ba8b59cc4f658f3af3327/lib/core.js#L544-L553

**Comment by goodmind** [comment]
Oh, I have PR for this https://github.com/facebook/flow/pull/7932

**Comment by goodmind** [comment]
Actually, [this](https://flow.org/try/#0MYewdgzgLgBAZgVzMGBeGAKAlALhgEgCUBTAQwBMB5MAGwE8BZUgBwB5oAnASzAHMAaAiQrV6AQQ4dSddlG58AfArTKA3gCgYMUJFgBbFniJkqtRi1nyBQk6LoSpMzj15K0MMMQDuMJs2wA3OqaMAD0odog5MQwUAAWpLA8XFBcpDRcAF7EELFxMQbMMKRg5DApuQBu6Qg5wVocxFAIHGAwhUEAvgFAA) works, so I don't think they're related

**Comment by goodmind** [comment]
nvm, they're related

**Comment by Bannerets** [comment]
@stekycz 

> `The original code looks like [...] `

Working as intended. `Map<any, string[]>` isn't a subtype of `Map<any, $ReadOnlyArray<string>>`. `$ReadOnlyMap<any, string[]>`, however, is.

Same as with these:

```js
class Inv<T> {}
class Co<+T> {}

;(new Inv<string> : Inv<string | number>) // error
;(new Co<string> : Co<string | number>) // ok
```

https://flow.org/en/docs/lang/variance/

**Comment by Bannerets** [comment]
The code in the original post is working as intended, the different behaviour is unsafe.

Flow won't show errors if you replace `class Parent<K> {}` with `class Parent<+K> {}`.

Same as with these:

```js
class Inv<T> {}
class Co<+T> {}

;(new Inv<string> : Inv<string | number>) // error
;(new Co<string> : Co<string | number>) // ok
```

https://flow.org/en/docs/lang/variance/
https://flow.org/en/docs/types/generics/#variance-sigils

**Comment by stekycz** [comment]
I understand that. However, it would be nice to have explained why the default behavior is invariant instead of expected and usual (at least from my point of view) covariance.
